modules/chapter1/pages/section2.adoc

```diff
@@ -1,3 +1,355 @@
+
+
+In more technical terms A *pipeline* is a description of an ML workflow, including all of the components in the workflow and how they combine in the form of a graph. The pipeline includes the definition of the inputs (parameters) required to run the pipeline and the inputs and outputs of each component.
+
+[NOTE]
+====
+ML Pipleines are a basically an execution graph of tasks, commonly known as a _DAG_ (Directed Acyclic Graph).
+A DAG is a directed graph with no cycles aka no direct loops
+====
+
+== Data Science Pipelines in OpenShift AI
+
+OpenShift AI offers two out of the box mechanisms to build Data Science Pipelines
+
+The first mechanism is the Elyra Jupyter notebook addon which provides a visual editor for creating pipelines via orchestration of Jupyter notebooks. 
+
+The second mechanism and the one discussed here is using Data Science Pipelines (DSP) based on the upstream KubeflowPipelines V1. With DSP, pipelines are python scripts built using the KFPV1 python SDK. Once built the python pipeline is submitted to the DSP runtime to be scheduled for execution.
+
+[NOTE]
+====
+There is work in progress to migrate DSP to Kubeflow Pipeline V2. At the time of writing this is still a work in progress and we won't cover it here.
+====
+
+In OpenShift AI we use the *_Tekton_* runtime to execute pipelines. With this there is an additional step to be performed the python script needs to be compiled into a Tekton definition before being submitting to the runtime
+
+
+In OpenShift AI the Data Science Pipeline runtime consists of the following components
+
+* A Data Science Pipeline Server container a number of resources 
+* A MariaDB for storing pipeline definitions and results
+* A Pipeline scheduler for scheduling pipeline runs
+* A Persistent Agent to record the set of containers that executed as well as their inputs and outputs.
+
+As pipelines are executed multiple ephemeral pods will be created and terminated depending on the execution steps of the pipelines
+
+[NOTE]
+====
+DS Pipeline in OpenShift AI are managed by the 'data-science-pipelines-operator-controller-manager' operator in the redhat-ods-applications namespace. The CRD is an instance of _datasciencepipelinesapplications.datasciencepipelinesapplications.opendatahub.io_
+====
+
+.DataScience Pipeline pods
+****
+[source,cmd]
+----
+ ~/oc get pods
+NAME                                                              READY   STATUS    RESTARTS   AGE
+ds-pipeline-persistenceagent-pipelines-definition-6d6676d58vzss   1/1     Running   0          9m49s
+ds-pipeline-pipelines-definition-7c7dd56b4d-qfz8t                 2/2     Running   0          9m49s
+ds-pipeline-scheduledworkflow-pipelines-definition-7f5c645hcg4g   1/1     Running   0          9m48s
+mariadb-pipelines-definition-795c57795b-27ljh                     1/1     Running   0          9m49s
+----
+****
+
+== Creating a Data Science Pipeline with KFP
```

Comment:
Clearly demarcate when the lecture narrative ends, and when the actual hands-on lab exercise begins - by prefixing the sub setion with Lab: or Exercise:. See rhods-intro repo for sample

Reply:
Also, if they have to create a workbench, let's tell them which one. I assume "Minimal Python" should be ok.